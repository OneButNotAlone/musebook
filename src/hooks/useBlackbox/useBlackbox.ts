import * as React from 'react';
import { useRef } from 'react';

/**
 * Blackbox container heavily inspired by Swizec Teller's teachings and [lib](https://github.com/Swizec/d3blackbox).
 * 
 * @param render a render prop that provides a ref you can bind to d3 with
 */
export const useBlackbox = function(render: (ref: HTMLElement) => void) {
    const refAnchor = React.useRef(null);

    React.useEffect(() => {
        render(refAnchor.current);
    });

    return refAnchor;
};
