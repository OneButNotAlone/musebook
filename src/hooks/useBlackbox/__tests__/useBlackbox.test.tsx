import * as React from 'react';
import { render } from 'react-testing-library';
import { useBlackbox } from '../useBlackbox';

test('blackbox render', () => {
    function renderThing(anchor: HTMLElement) {
        anchor.append('div');
    }
    const ThingComponent = ({ x, y }: { x: number; y: number }) => {
        const refAnchor = useBlackbox((anchor: HTMLElement) => renderThing(anchor));

        return (
            <>
                <g ref={refAnchor} transform={`translate(${x}, ${y})`} />
            </>
        );
    };

    const { container } = render(<ThingComponent x={0} y={0} />);

    expect(container).toMatchSnapshot();
});
