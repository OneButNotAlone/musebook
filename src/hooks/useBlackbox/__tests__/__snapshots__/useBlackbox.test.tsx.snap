// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blackbox render 1`] = `
<div>
  <g
    transform="translate(0, 0)"
  >
    div
  </g>
</div>
`;
